include/envoy/network/listen_socket.h

```diff
@@ -15,9 +16,31 @@
 namespace Envoy {
 namespace Network {
 
-// Optional variant of setsockopt(2) optname. The idea here is that if the option is not supported
-// on a platform, we can make this the empty value. This allows us to avoid proliferation of #ifdef.
-using SocketOptionName = absl::optional<std::pair<int, int>>;
+// SocketOptionName is an optional value that captures the setsockopt(2)
+// arguments. The idea here is that if a socket option is not supported
+// on a platform, we can make this the empty value, which allows us to
+// avoid #ifdef proliferation of #ifdef.
+struct SocketOptionName {
+  SocketOptionName() = default;
+  SocketOptionName(const SocketOptionName&) = default;
+  SocketOptionName(int level, int option, const std::string& name)
+      : value_(std::make_tuple(level, option, name)) {}
+
+  int level() const { return std::get<0>(value_.value()); }
+  int option() const { return std::get<1>(value_.value()); }
+  const std::string& name() const { return std::get<2>(value_.value()); }
+
+  bool has_value() const { return value_.has_value(); }
+  bool operator==(const SocketOptionName& rhs) const { return value_ == rhs.value_; }
+
+private:
+  absl::optional<std::tuple<int, int, std::string>> value_;
```

Comment:
nit: const

Reply:
```
source/common/network/socket_option_impl.cc:41:14: error: object of type 'Envoy::Network::SocketOptionName' cannot be assigned because its copy assignment operator is implicitly deleted
  info.name_ = optname_;
```

If I make this const, then I'd also need to make assignment work, since assigning this struct seems like a reasonable thing to support. I'd prefer to leave it non-const.

Reply:
OK sounds good.